ofrak_core/ofrak/gui/script_builder.py

```diff
@@ -0,0 +1,348 @@
+from dataclasses import dataclass, field
+from enum import IntEnum
+import re
+import os
+from typing import Dict, List, Optional, Tuple
+from ofrak.model.resource_model import ResourceIndexedAttribute
+from ofrak.core.filesystem import FilesystemEntry
+from ofrak.model.resource_model import Data
+from ofrak.service.resource_service_i import ResourceAttributeValueFilter, ResourceFilter
+from black import format_str, FileMode
+
+from ofrak.resource import Resource
+
+
+class SelectableAttributesError(Exception):
+    """
+    Prompt the user for an attribute to select with
+    """
+
+
+class ActionType(IntEnum):
+    UNPACK = 0
+    MOD = 1
+    PACK = 2
+    UNDEF = 3
+
+
+@dataclass(frozen=True)
+class ScriptAction:
+    """
+    Encapsulates the structure of a single action within the script, which consists of the string
+    representation of the code for that action and the action's type.
+    """
+
+    action_type: ActionType
+    action: str
+
+
+@dataclass
+class ScriptSession:
+    """
+    A script, consisting of an ordered sequence of script actions and a mapping between resources
+    and their autogenerated variable names.
+    """
+
+    actions: List[ScriptAction] = field(default_factory=list)
+    resource_variable_names: Dict[bytes, str] = field(default_factory=dict)
+    boilerplate_header: str = r"""
+    from ofrak import *
+    from ofrak.core import *
+
+    async def main(ofrak_context: OFRAKContext):"""
+    # TODO: Replace with backend in use by OFRAK instance used to create the script.
+    boilerplate_footer: str = r"""
+    if __name__ == "__main__":
+        if False:
+            import ofrak_angr
+            import ofrak_capstone
+            ofrak.discover(ofrak_capstone)
+            ofrak.discover(ofrak_angr)
+
+        if False:
+            import ofrak_binary_ninja
+            import ofrak_capstone
+            ofrak.discover(ofrak_capstone)
+            ofrak.discover(ofrak_binary_ninja)
+
+        if False:
+            import ofrak_ghidra
+            ofrak.discover(ofrak_ghidra)
+
+        ofrak = OFRAK()
+        ofrak.run(main)
+    """
+
+
+class ScriptBuilder:
+    """
+    Builds and maintains runnable OFRAK scripts as sequences of actions, with each script tied to
+    a session.
+    """
+
+    def __init__(self):
+        self.root_cache: Dict[bytes, Resource] = {}
+        self.script_sessions: Dict[bytes, ScriptSession] = {}
+        self.selectable_indexes: List[ResourceIndexedAttribute] = [
+            FilesystemEntry.Name,
+            Data.Offset,
+        ]
```

Comment:
I think we should try using virtual addresses as identifiers before we use offsets. Not only will virtual addresses be unique more than offsets will, but they will make the generated code more closely correspond with the displayed info in the GUI. 

For example, without this, a complex block in the GUI might be shown as: `0xd80: main`, but in the generated code it is `complexblock_0x2e0` because that is its offset. I find this confusing, and think using virtual addresses when possible would address it.

```suggestion
        self.selectable_indexes: List[ResourceIndexedAttribute] = [
            FilesystemEntry.Name,
            Addressable.VirtualAddress,
            Data.Offset,
        ]
```